// sol-merger/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
}
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolidityImportVisitor = exports.SolidityExportVisitor = exports.ImportsAnalyzer = exports.ExportsAnalyzer = exports.merge = exports.FileAnalyzer = exports.Merger = void 0;
const exportVisitor_1 = require("./antlr/visitors/exportVisitor");
Object.defineProperty(exports, "SolidityExportVisitor", { enumerable: true, get: function () { return exportVisitor_1.SolidityExportVisitor; } });
const importVisitor_1 = require("./antlr/visitors/importVisitor");
Object.defineProperty(exports, "SolidityImportVisitor", { enumerable: true, get: function () { return importVisitor_1.SolidityImportVisitor; } });
const exportsAnalyzer_1 = require("./exportsAnalyzer");
Object.defineProperty(exports, "ExportsAnalyzer", { enumerable: true, get: function () { return exportsAnalyzer_1.ExportsAnalyzer; } });
const fileAnalyzer_1 = require("./fileAnalyzer");
Object.defineProperty(exports, "FileAnalyzer", { enumerable: true, get: function () { return fileAnalyzer_1.FileAnalyzer; } });
const importsAnalyzer_1 = require("./importsAnalyzer");
Object.defineProperty(exports, "ImportsAnalyzer", { enumerable: true, get: function () { return importsAnalyzer_1.ImportsAnalyzer; } });
const merger_1 = require("./merger");
Object.defineProperty(exports, "Merger", { enumerable: true, get: function () { return merger_1.Merger; } });
const merge = (file, options = {}) => __awaiter(void 0, void 0, void 0, function* () {
    const merger = new merger_1.Merger(options);
    return merger.processFile(file, true);
});
exports.merge = merge;
__exportStar(require("./antlr/visitors/types"), exports);
var plugins_1 = require("./plugins");
Object.defineProperty(exports, "plugins", { enumerable: true, get: function () { return plugins_1.plugins; } });
